src/osp/Active/SysSceneGraph.h

```diff
@@ -0,0 +1,188 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#pragma once
+
+#include "basic.h"
+#include "activetypes.h"
+
+#include <algorithm>
+
+namespace osp::active
+{
+
+/**
+ * @brief Helps add entities to a reserved space in ACtxSceneGraph
+ *
+ * All reserved space must be used, or else this class will assert on
+ * destruction.
+ */
+class SubtreeBuilder
+{
+public:
+    constexpr SubtreeBuilder(ACtxSceneGraph& rScnGraph, ActiveEnt root, TreePos_t first, TreePos_t last)
+     : m_rScnGraph{rScnGraph}
+     , m_root{root}
+     , m_first{first}
+     , m_last{last}
+    { }
+    constexpr SubtreeBuilder(SubtreeBuilder&& move) = default;
+    SubtreeBuilder(SubtreeBuilder const& copy) = delete;
+    ~SubtreeBuilder() { assert(m_first == m_last); }
+
+    /**
+     * @brief Add a child that might have more descendants
+     *
+     * @return
+     */
+    [[nodiscard]] SubtreeBuilder add_child(ActiveEnt ent, uint32_t descendantCount);
+
+    void add_child(ActiveEnt ent)
+    {
+        (void) add_child(ent, 0);
+    }
+
+    std::size_t remaining()
+    {
+        return m_last - m_first;
+    }
+
+private:
+    ActiveEnt m_root;
+    TreePos_t m_first;
+    TreePos_t m_last;
+
+    ACtxSceneGraph& m_rScnGraph;
+
+}; // class SubtreeBuilder
+
+/**
+ * @brief Iterates entity children in an ACtxSceneGraph
+ */
+class ChildIterator
+{
+public:
+    using iterator_category = std::forward_iterator_tag;
+    using difference_type   = std::ptrdiff_t;
+    using value_type        = ActiveEnt;
+    using pointer           = void;
+    using reference         = void;
+
+    constexpr ChildIterator() noexcept = default;
+    constexpr ChildIterator(ACtxSceneGraph const* pScnGraph, TreePos_t pos) noexcept
+     : m_pScnGraph{pScnGraph}
+     , m_pos{pos}
+    { }
+    constexpr ChildIterator(ChildIterator const& copy) noexcept = default;
+    constexpr ChildIterator(ChildIterator&& move) noexcept = default;
+
+    constexpr ChildIterator& operator=(ChildIterator const& copy) noexcept = default;
+    constexpr ChildIterator& operator=(ChildIterator&& move) noexcept = default;
+
+    ChildIterator& operator++() noexcept
+    {
+        m_pos += 1 + m_pScnGraph->m_treeDescendants[m_pos];
+        return *this;
+    }
+
+    auto operator<=>(ChildIterator const&) const = default;
```

Comment:
Minor: My implementation experience with operator<=> at work makes me recommend uisng
```
friend auto operator<=>(ChildIterator const& lhs,ChildIterator const& rhs) noexcept = default;
```

Something about the way the operator<=> auto-generates things causes some hard to comprehend compiler errors in rare situations that using "friend" makes go away.